internal/cloud/backend.go

```diff
@@ -941,6 +949,34 @@ func (wm WorkspaceMapping) Strategy() workspaceStrategy {
 	}
 }
 
+func (wm WorkspaceMapping) tfeTags() []*tfe.Tag {
+	var tags []*tfe.Tag
+
+	for _, tag := range wm.Tags {
+		t := tfe.Tag{Name: tag}
+		tags = append(tags, &t)
+	}
+
+	return tags
+}
+
+func (wm WorkspaceMapping) newTags(workspaceTags []string) []*tfe.Tag {
+	wsTags := map[string]struct{}{}
+	for _, t := range workspaceTags {
+		wsTags[t] = struct{}{}
+	}
+
+	var tags []*tfe.Tag
+	for _, tag := range wm.tfeTags() {
+		if _, ok := wsTags[tag.Name]; !ok {
+			t := tfe.Tag{Name: tag.Name}
+			tags = append(tags, &t)
+		}
+	}
+
+	return tags
+}
```

Comment:
Attempting to diff the tags like this seems like it's prone to a read-modify-write concurrency error. What would happen if instead we just call the `AddTags` method passing the entire tag set? Would Terraform Cloud just ignore any already-set tags? If so, that would avoid the RMW problem.

Reply:
I think duplicated tag values would be ignored, yes. I don't believe we need to do this sort of diffing, to determine what tags are missing and explicitly added but [I mentioned here](https://github.com/hashicorp/terraform/pull/29853#discussion_r740710308) that I don't think we should make a call out to tags every time `StateMgr()` is called; what do you think of that? 

Would "Loop through existing tags, if any are missing, send the full tag set" alleviate this concern?

Reply:
(I've updated the PR to reflect what I meant here)

Reply:
Yep, that looks good. The issue is quite academic anyway, so it's more about clarity than any practical concern, and the updated code reads well to me.